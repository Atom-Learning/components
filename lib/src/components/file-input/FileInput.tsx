import { Upload } from '@atom-learning/icons'
import * as React from 'react'

import { Button } from '../button'
import { Icon } from '../icon'

export type FileInputProps = React.ComponentProps<typeof Button> & {
  onFileSelect: (selection: FileList | null) => void
  accept?: string
  multiple?: boolean
}

export const FileInput: React.FC<FileInputProps> = ({
  accept,
  children,
  multiple = false,
  onFileSelect,
  ...rest
}) => {
  const handleFileSelect = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { files } = event.target

    onFileSelect(files)
  }

  return (
    <Button {...rest} asChild>
      <label>
        <Icon is={Upload} />
        {children}
        <input
          type="file"
          onChange={handleFileSelect}
          accept={accept}
          multiple={multiple}
          hidden
        />
      </label>
    </Button>
  )
}

FileInput.displayName = 'FileInput'
