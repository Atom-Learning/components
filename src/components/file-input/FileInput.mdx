---
title: File Input
component: FileInput
description: FileInput is the lowest-level input component for the file type
category: Form primitives
---

`FileInput` is a specialised input that allows file uploads.

The 3 most important props it takes are: `onSelect`, `id` and `label`.

`onSelect` is a callback that accepts the actual file selection (as a FileList), which gets updated after the user interacts with the file browser interface. The file selection can then be used in order to build a payload for something like a `POST` API call.

Note: The actual action of uploading the file to a backend is the parent's responsibility.

`id` is really important because the actual file input is hidden, while the visible element is a `label` with an `htmlFor` attribute that pairs with the hidden input via the `id` attribute. 

 Note: `id` has to be unique 

`label` is used to add text to the visible button.

```jsx preview
<FileInput
  onSelect={(fileSelection) => console.log(fileSelection)}
  id="unique-identifier"
  label="Select file"
/>
```

The input can also accept multiple files at once, via the `multiple` prop. Another important feature is the `accept` prop. This makes the input filter out unwanted file types and prevent the user from selecting them.

```jsx preview
<FileInput
  onSelect={(fileSelection) => console.log(fileSelection)}
  id="unique-identifier"
  label="Select file"
  multiple
  accept="application/pdf"
/>
```

You can also use our icon pack to add an icon next to the button text.

```jsx preview
<FileInput
  onSelect={(fileSelection) => console.log(fileSelection)}
  id="unique-identifier"
  label="Select file"
  multiple
  accept="application/pdf"
  icon={Upload}
/>
```
