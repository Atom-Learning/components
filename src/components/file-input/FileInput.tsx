import { Download } from '@atom-learning/icons'
import { CSS } from '@stitches/react'
import * as React from 'react'

import { Button } from '../button'
import { Icon } from '../icon'

export type FileInputProps = {
  onSelect: (selection: FileList) => void
  id: string
  label: string
  accept?: string
  multiple?: boolean
  appearance?: string
  css?: CSS
  icon?: SVGSVGElement
}

export const FileInput: React.FC<FileInputProps> = ({
  onSelect,
  label,
  accept,
  multiple,
  id,
  appearance = 'outline',
  icon,
  css
}) => {
  const [fileSelection, setFileSelection] = React.useState<FileList>()

  React.useEffect(() => {
    onSelect(fileSelection)
  }, [fileSelection, onSelect])

  const handleFileSelect = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFileSelection(event.target.files)
  }

  return (
    <>
      <input
        type="file"
        onChange={handleFileSelect}
        accept={accept}
        multiple={multiple}
        id={id}
        data-testid="file-upload"
        hidden
      />
      <Button as="label" htmlFor={id} appearance={appearance} css={css}>
        <Icon is={Download} />
        {label}
      </Button>
    </>
  )
}

FileInput.displayName = 'FileInput'
