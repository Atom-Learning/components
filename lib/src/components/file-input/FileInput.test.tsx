import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { axe } from 'jest-axe'
import * as React from 'react'
import { vi } from 'vitest'

import { FileInput } from '.'
import { FileInputProps } from './FileInput'

describe('FileInput component', () => {
  const props: FileInputProps = {
    onFileSelect: vi.fn()
  }

  it('renders', async () => {
    const { container } = render(<FileInput {...props}>Upload</FileInput>)

    expect(container).toMatchSnapshot()
  })

  it('has no programmatically detectable a11y issues', async () => {
    const { container } = render(<FileInput {...props}>Upload</FileInput>)

    expect(await axe(container)).toHaveNoViolations()
  })

  it('uploads a file', async () => {
    const testFile = new File(['random content'], 'upload.txt')
    render(<FileInput {...props}>Upload</FileInput>)

    const input = screen.getByLabelText('Upload') as HTMLInputElement

    await waitFor(() =>
      fireEvent.change(input, {
        target: { files: [testFile] }
      })
    )

    expect(input.files[0].name).toBe('upload.txt')
    expect(input.files.length).toBe(1)
  })

  it('uploads multiple file', async () => {
    const testFile = new File(['random content'], 'upload.txt')
    const testFile2 = new File(['random content'], 'upload2.txt')
    render(
      <FileInput {...props} multiple>
        Upload
      </FileInput>
    )

    const input = screen.getByLabelText('Upload') as HTMLInputElement

    await waitFor(() =>
      fireEvent.change(input, {
        target: { files: [testFile, testFile2] }
      })
    )

    expect(input.files[0].name).toBe('upload.txt')
    expect(input.files[1].name).toBe('upload2.txt')
    expect(input.files.length).toBe(2)
  })
})
