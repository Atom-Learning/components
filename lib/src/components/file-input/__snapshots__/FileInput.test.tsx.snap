// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput component renders 1`] = `
@media  {
  .c-fkUJsw {
    align-items: center;
    background: unset;
    border: unset;
    border-radius: var(--radii-0);
    cursor: pointer;
    display: flex;
    font-family: var(--fonts-body);
    font-weight: 600;
    justify-content: center;
    padding: unset;
    text-decoration: none;
    transition: all 100ms ease-out;
    white-space: nowrap;
    width: max-content;
  }

  .c-dbrbZt {
    display: inline-block;
    fill: none;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }
}

@media  {
  .c-fkUJsw-dXsdwv-size-md {
    font-size: var(--fontSizes-md);
    line-height: 1.5;
    height: var(--sizes-4);
    padding-left: var(--space-5);
    padding-right: var(--space-5);
    gap: var(--space-3);
  }

  .c-fkUJsw-dXsdwv-size-md .c-dbrbZt {
    height: 20px;
    width: 20px;
  }

  .c-dbrbZt-dMrjnF-size-md {
    height: var(--sizes-2);
    width: var(--sizes-2);
    stroke-width: 1.75;
  }
}

@media  {
  .c-fkUJsw-gsycxF-cv {
    background: var(--colors-primary);
    color: white;
  }

  .c-fkUJsw-gsycxF-cv[disabled] {
    background: var(--colors-tonal100);
    color: var(--colors-tonal400);
    cursor: not-allowed;
  }

  .c-fkUJsw-gsycxF-cv:not([disabled]):hover,
  .c-fkUJsw-gsycxF-cv:not([disabled]):focus {
    background: var(--colors-primaryMid);
    color: white;
  }

  .c-fkUJsw-gsycxF-cv:not([disabled]):active {
    background: var(--colors-primaryDark);
  }
}

<div>
  <label
    class="c-fkUJsw c-fkUJsw-dXsdwv-size-md c-fkUJsw-gsycxF-cv"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="c-dbrbZt c-dbrbZt-dMrjnF-size-md"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 4v13"
      />
      <polyline
        points="7 8 12 3 17 8"
      />
      <path
        d="M20 21H4"
      />
    </svg>
    Upload
    <input
      hidden=""
      type="file"
    />
  </label>
</div>
`;
